platform/ui/src/components/AllInOneMenu/Menu.tsx

```diff
@@ -0,0 +1,145 @@
+import React, { createContext, ReactNode, useCallback, useEffect, useState } from 'react';
+import { useTranslation } from 'react-i18next';
+
+import './allInOneMenu.css';
+import DividerItem from './DividerItem';
+import PanelSelector from './PanelSelector';
+import classNames from 'classnames';
+import Icon from '../Icon';
+
+export interface MenuProps {
+  className?: string;
+  isVisible?: boolean;
+  preventHideMenu?: boolean;
+  backLabel: string;
+  headerComponent?: ReactNode;
+  showHeaderDivider?: boolean;
+  activePanelIndex?: number;
+  onMenuVisibilityChange?: (isVisible: boolean) => void;
+  children: unknown;
+}
+
+type MenuContextProps = {
+  showSubMenu: (subMenuProps: MenuProps) => void;
+  hideMenu: () => void;
+  addItemPanel: (index: number, label: string) => void;
+  activePanelIndex: number;
+};
+
+type MenuPathState = {
+  props: MenuProps;
+  activePanelIndex: number;
+};
+
+export const MenuContext = createContext<MenuContextProps>(null);
+
+const Menu = (props: MenuProps) => {
+  const { isVisible, onMenuVisibilityChange, activePanelIndex, preventHideMenu, className } = props;
+
+  const [isMenuVisible, setIsMenuVisible] = useState(false);
+  const [menuPath, setMenuPath] = useState<Array<MenuPathState>>([
+    { props, activePanelIndex: activePanelIndex || 0 },
+  ]);
+  const [itemPanelLabels, setItemPanelLabels] = useState<Array<string>>([]);
+
+  useEffect(() => {
+    setIsMenuVisible(isVisible);
+    onMenuVisibilityChange?.(isVisible);
+  }, [isVisible, onMenuVisibilityChange]);
+
+  const showSubMenu = useCallback((subMenuProps: MenuProps) => {
+    setMenuPath(path => {
+      return [
+        ...path,
+        { props: subMenuProps, activePanelIndex: subMenuProps.activePanelIndex || 0 },
+      ];
+    });
+    setItemPanelLabels([]);
+  }, []);
+
+  const hideMenu = useCallback(() => {
+    if (preventHideMenu) {
+      return;
+    }
+
+    setMenuPath(path => [path[0]]);
+    setItemPanelLabels([]);
+    setIsMenuVisible(false);
+    onMenuVisibilityChange(false);
+  }, [preventHideMenu, onMenuVisibilityChange]);
+
+  const addItemPanel = useCallback((index, label) => {
+    setItemPanelLabels(labels => {
+      labels.splice(index, 0, label);
+      return [...labels];
+    });
```

Comment:
probably wrong 

should be creating a new `updatedLabel = []` instead of mutating the state

Reply:
If I were mutating the state OUTSIDE of the state update function (i.e. in this case setItemPanelLabels) and passing the mutated state to the update function then that would be wrong. But instead here I am passing a function to the update function that is returning a new, updated state. See here https://www.codingdeft.com/posts/react-usestate-array/#modifying-an-item-in-the-array

Reply:
But I suppose I can clean this up a bit.